plugins/inputs/intel_powerstat/intel_powerstat.go

```diff
@@ -1,915 +1,1218 @@
-	msr  msrService
-
-	cpuFrequency        bool
-	cpuBusyFrequency    bool
-	cpuTemperature      bool
-	cpuC0StateResidency bool
-	cpuC1StateResidency bool
-	cpuC6StateResidency bool
-	cpuBusyCycles       bool
-
-	packageTurboLimit                  bool
-	packageCurrentPowerConsumption     bool
-	packageCurrentDramPowerConsumption bool
-	packageThermalDesignPower          bool
-	packageUncoreFrequency             bool
-	packageCPUBaseFrequency            bool
-
-	cpuBusClockValue   float64
-	cpuInfo            map[string]*cpuInfo
-	skipFirstIteration bool
-	logOnce            map[string]error
+	CPUMetrics       []string        `toml:"cpu_metrics"`
+	PackageMetrics   []string        `toml:"package_metrics"`
+	IncludedCPUs     []string        `toml:"included_cpus"`
+	ExcludedCPUs     []string        `toml:"excluded_cpus"`
+	EventDefinitions string          `toml:"event_definitions"`
+	MsrReadTimeout   config.Duration `toml:"msr_read_timeout"`
+	Log              telegraf.Logger `toml:"-"`
+
+	parsedIncludedCores []int
+	parsedExcludedCores []int
+
+	parsedCPUTimedMsrMetrics []string
+	parsedCPUPerfMetrics     []string
+	parsedPackageRaplMetrics []string
+	parsedPackageMsrMetrics  []string
+
+	option  OptGenerator
+	fetcher MetricFetcher
+
+	cpuMetricsMap    map[string]func(telegraf.Accumulator, int, int, int)
+	packageMetricMap map[string]func(telegraf.Accumulator, int)
+
+	needsCoreFreq       bool
+	needsMsrCPU         bool
+	needsPerf           bool
+	needsTimeRelatedMsr bool
+
+	needsRapl       bool
+	needsMsrPackage bool
+
+	logOnce map[string]struct{}
 }
 
+// SampleConfig returns a sample configuration (See sample.conf).
 func (*PowerStat) SampleConfig() string {
 	return sampleConfig
 }
 
-// Init performs one time setup of the plugin
-func (p *PowerStat) Init() error {
-	p.parsePackageMetricsConfig()
-	p.parseCPUMetricsConfig()
-	err := p.verifyProcessor()
-	if err != nil {
+// Start initializes the MetricFetcher interface of the receiver to gather metrics.
+func (p *PowerStat) Start(_ telegraf.Accumulator) error {
+	if err := p.disableUnsupportedMetrics(); err != nil {
+		return err
+	}
+
+	if err := p.parseConfig(); err != nil {
 		return err
 	}
 
-	p.initMSR()
-	p.initRaplService()
+	opts := p.option.Generate(OptConfig{
+		cpuMetrics:     p.CPUMetrics,
+		packageMetrics: p.PackageMetrics,
+		includedCPUs:   p.parsedIncludedCores,
+		excludedCPUs:   p.parsedExcludedCores,
+		perfEventFile:  p.EventDefinitions,
+		msrReadTimeout: time.Duration(p.MsrReadTimeout),
+		log:            p.Log,
+	})
 
-	if !p.areCoreMetricsEnabled() && !p.areGlobalMetricsEnabled() {
-		return fmt.Errorf("all configuration options are empty or invalid. Did not find anything to gather")
+	var err error
+	var initErr *ptel.MultiError
+	p.fetcher, err = ptel.New(opts...)
+	if err != nil {
+		if !errors.As(err, &initErr) {
+			return fmt.Errorf("failed to initialize metric fetcher interface: %w", err)
+		}
+		p.Log.Warnf("Plugin initialized with errors: %v", err)
```

Comment:
The reason is `MultiError` is a custom type error used to mark when a specific module fails to initialize, and hence the metrics that rely on these modules will not be gathered. The idea was to let the plugin continue (best effort) providing as many metrics as it can, and provide the user with the error so that user knows the reason why few metrics might not be displayed.

Reply:
While I can accept that behavior I would like you to consider that many people do not read log-files until they don't get any metric. :-) Anyway, a short comment in the code would be nice as to why you _can_ and _want_ to continue.

Reply:
```suggestion
		p.Log.Warnf("Plugin started with errors: %v", err)
```

Since we are in the `Start()` method, the plugin is already initialised.